agent/hcp/manager.go

```diff
@@ -108,6 +112,12 @@ func (m *Manager) Run(ctx context.Context) {
 
 	// main loop
 	for {
+		// Check for configured management token from HCP. It MUST NOT override the user-provided initial management token.
```

Comment:
This comment reminds me, are we safeguarding from calling this initialize management token method before the user-provided token initialization? This is the scenario I'm thinking of:

1. User is bootstrapping their server (i.e., creating and not linking an existing one) and has both `cloud` and [`initial_management`](https://developer.hashicorp.com/consul/docs/agent/config/config-files#acl_tokens_initial_management) configured.
2. Node becomes the leader, `establishLeadership` > `initializeACLs` > `initializeManagementToken` is called with the initial_management token.
3. In parallel, the HCP manager is started and calls `initializeManagementToken` with the HCP token.

My hope is that the HCP manager's `initializeManagementToken` would fail if `establishLeadership` hasn't completed, but not sure about the behavior here.

Reply:
It looks like `initializeManagementToken` can do one of two things. First if ACLs havn't been bootstrapped (checking a sentinel value in state store) then it initiates bootstrapping using the given secretID as the root token. If ACLs have been bootstrapped then the token just gets upserted with the global management policy keying off the given secretID. The bootstrapping only happens when the first leadership is established. Future changes in leadership should trigger the upsert logic since the bootstrap sentinel value exists in the state store.

I think the scenario described (HCP management token gets bootstrapped before the user provided initial management token) is possible. I also think there's an inverse scenario possible when a user doesn't set an initial management token (e.g. all HCP managed clusters) where `initializeACLs` fails to write the anonymous token since ACLs havent yet been bootstrapped (by hcp manager calling initializeManagementToken).

If linking via the API a user will already have ACLs bootstrapped so I don't think we need the bootstrapping behavior. Maybe we can seperate out these two different fuctions of the `initializeManagementToken` function and just hand the hcp manager a callback to upsert a token with the global management policy (e.g. a management token). I think this also means we cannot get away from having the check for a file provided value in the `server.initializeACLs` since it will need to be known in order to bootstrap ACLs for HCP managed clusters.

Reply:
> I also think there's an inverse scenario possible when a user doesn't set an initial management token (e.g. all HCP managed clusters) where initializeACLs fails to write the anonymous token since ACLs havent yet been bootstrapped (by hcp manager calling initializeManagementToken).

From what I can tell, this doesn't _seem_ to be an issue unless I'm misunderstanding. The anonymous token is able to be created even if no token is bootstrapped before it. I can see this if I don't set an initial management token and then step through the code with a debugger — the anonymous token gets created with no error.

That being said, I'm all for staying safe and keeping the existing `initializeManagementToken` call in place and just doing a simplified "upsert" in the HCP manager.

Reply:
I've changed this so the HCP manager now calls `upsertManagementToken` which does as you mentioned — it does not try to bootstrap and only tries to upsert the token.

I've also put the `initializeManagementToken` call for the "HCP Management Token" back into the `establishLeadership > initializeACLs` process.

It _seems_ like this should work for all scenarios, but I'm not totally sure. There is a possibility that `upsertManagementToken` in HCP manager can be called before, during, or after `establishLeadership` finishes. 

If it runs **before**, the `initializeManagementToken` call for the user-provided token will be properly bootstrapped and the `initializeManagementToken` call for the "HCP Management Token" will be a no-op.

If it runs **after** `establishLeadership` finishes, it is a no-op.

If it runs **during** `establishLeadership`, there is a possibility that both the HCP manager and the `establishLeadership` code try to create the "HCP Management Token" at the same time. I'm not totally sure if this is dangerous. Any thoughts?